Explain the non-transparent mode design decision for TCP/UDP (#685)
envoyproxy/gateway: docs/latest/design/tcp-udp-design.md

> # TCP&UDP Proxy Design 

Reviewer: ```suggestion
# TCP and UDP Proxy Design 
```

Even though most of the use cases for Envoy Gateway are at Layer-7, Envoy Gateway can also work at Layer-4 to proxy TCP 
and UDP traffic. This document will explore the options we have when operating Envoy Gateway at Layer-4 and explain the 
design decision.

> Envoy can work as a non-transparent proxy or a [transparent proxy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_features/ip_transparency#arch-overview-ip-transparency-original-src-listener), 

Reviewer: the links for UDP transparant/non transparent modes are different from TCP , that should be highlighted here

Author: The envoy documentation is a bit confusing, I though [Original Source Listener Filter] works for both TCP and UDP.

so ideally, Envoy Gateway should also be able to work in these two modes:

> ## Non-transparent Proxy Mode

Reviewer: if you are clubbing TCP and UDP as one, then you can't use `terminate` connection since its doesn't apply to UDP :)
so recommend either creating two separate sections for UDP and TCP

Envoy terminates the downstream connection, connects the upstream with its own address,  and proxies the TCP/UDP traffic
from the downstream to the upstream. In this mode, the upstream will see Envoy's address.

## Transparent Proxy Mode
Envoy terminates the downstream connection, connects the upstream with the downstream source address,  and proxies the
TCP/UDP traffic from the downstream to the upstream. In this mode, the upstream will see the original downstream 
address.

## The Implications of Transparent Proxy Mode

### Escalated Privilege
Envoy needs to bind to the downstream source IP when connecting to the upstream, which means Envoy requires 
escalated CAP_NET_ADMIN privileges. This is often considered as a bad security practice and not allowed in some 
sensitive deployments.

### Routing
The upstream can see the original source IP, but the original port number won't be past through, so the return 
traffic from the upstream must be routed back to Envoy because only Envoy knows how to send the return traffic back 

> to the right port number of the downstream, which requires routing at the upstream side must be deliberately set up. 

Reviewer: ```suggestion
to the right port number of the downstream, which requires routing at the upstream side to be set up. 
```

In a Kubernetes cluster, Envoy Gateway will have to carefully cooperate with CNI plugins to get the routing right.

> ## The Design Decision(For Now)

Reviewer: ```suggestion
## The Design Decision (For Now)
```

> Since now Envoy Gateway's primary goal is implementing an in-cluster Gateway, and it's tricky to support non-transparent 

Reviewer: you can reuse some of the language from https://github.com/envoyproxy/gateway/pull/673 and we can close that issue in favor of this one

design decision.

Envoy can work as a non-transparent proxy or a transparent proxy for both [TCP](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_features/ip_transparency#arch-overview-ip-transparency-original-src-listener)
 and [UDP](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto#envoy-v3-api-msg-extensions-filters-udp-udp-proxy-v3-udpproxyconfig)
, so ideally, Envoy Gateway should also be able to work in these two modes:

## Non-transparent Proxy Mode
For TCP, Envoy terminates the downstream connection, connects the upstream with its own IP address, and proxies the 
TCP traffic from the downstream to the upstream. 

For UDP, Envoy receives UDP packages from the downstream, and use its own IP address as the sender IP address when 
proxying the UDP packages to the upstream.

In this mode, the upstream will see Envoy's IP address.

## Transparent Proxy Mode
For TCP, Envoy terminates the downstream connection, connects the upstream with the downstream IP address, and proxies 
the TCP traffic from the downstream to the upstream. 

> For UDP, Envoy receives UDP packages from the downstream, and use the downstream IP address as the sender IP address 

Reviewer: ```suggestion
For UDP, Envoy receives UDP packages from the downstream, and uses the downstream IP address as the sender IP address 
```

when proxying the UDP packages to the upstream.

In this mode, the upstream will see the original downstream IP address.

Note: Even in transparent mode, the upstream can't see the port number of the downstream because Envoy doesn't forward 
the port number.

## The Implications of Transparent Proxy Mode

### Escalated Privilege
Envoy needs to bind to the downstream IP when connecting to the upstream, which means Envoy requires escalated 
CAP_NET_ADMIN privileges. This is often considered as a bad security practice and not allowed in some sensitive deployments.

### Routing

> The upstream can see the original source IP, but the original port number won't be past through, so the return 

Reviewer: ```suggestion
The upstream can see the original source IP, but the original port number won't be passed through, so the return 
```

> The upstream can see the original source IP, but the original port number won't be past through, so the return 

Reviewer: ```suggestion
The upstream can see the original source IP, but the original port number won't be passed, so the return 
```

traffic from the upstream must be routed back to Envoy because only Envoy knows how to send the return traffic back
to the right port number of the downstream, which requires routing at the upstream side to be set up. 

> In a Kubernetes cluster, Envoy Gateway will have to carefully cooperate with CNI plugins to get the routing right.

Reviewer 2: Can you create and link an issue that will be used to capture the details of integrating transparent proxy mode with the kube network.
